# Амиров Дмитрий, РИ-400024, HW_1

- Импортирование необходимых библиотек

In [49]:
# C:\Users\amirov\AppData\Local\Programs\Python\Python312\Scripts\pip.exe install polars pandas 

import polars as pl
import pandas as pd
import bottleneck as bn

# Работа с Polars

- Чтение датасета с помощью polars

In [50]:
df_amirov = pl.read_csv("train.csv")

df_amirov.head(5)

shape: (5, 12)
┌─────────────┬──────────┬────────┬───────────────────┬───┬───────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name              ┆ … ┆ Ticket    ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---               ┆   ┆ ---       ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str               ┆   ┆ str       ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪═══════════════════╪═══╪═══════════╪═════════╪═══════╪══════════╡
│ 1           ┆ 0        ┆ 3      ┆ Braund, Mr. Owen  ┆ … ┆ A/5 21171 ┆ 7.25    ┆ null  ┆ S        │
│             ┆          ┆        ┆ Harris            ┆   ┆           ┆         ┆       ┆          │
│ 2           ┆ 1        ┆ 1      ┆ Cumings, Mrs.     ┆ … ┆ PC 17599  ┆ 71.2833 ┆ C85   ┆ C        │
│             ┆          ┆        ┆ John Bradley      ┆   ┆           ┆         ┆       ┆          │
│             ┆          ┆        ┆ (Flor…            ┆   ┆           ┆         ┆       ┆          │
│ 3           ┆ 1        ┆ 3      ┆ Heikkinen, Miss.  ┆ … ┆ STON/O2.  ┆ 7.925   ┆ null  ┆ S        │
│             ┆          ┆        ┆ Laina             ┆   ┆ 3101282   ┆         ┆       ┆          │
│ 4           ┆ 1        ┆ 1      ┆ Futrelle, Mrs.    ┆ … ┆ 113803    ┆ 53.1    ┆ C123  ┆ S        │
│             ┆          ┆        ┆ Jacques Heath     ┆   ┆           ┆         ┆       ┆          │
│             ┆          ┆        ┆ (Li…              ┆   ┆           ┆         ┆       ┆          │
│ 5           ┆ 0        ┆ 3      ┆ Allen, Mr.        ┆ … ┆ 373450    ┆ 8.05    ┆ null  ┆ S        │
│             ┆          ┆        ┆ William Henry     ┆   ┆           ┆         ┆       ┆          │
└─────────────┴──────────┴────────┴───────────────────┴───┴───────────┴─────────┴───────┴──────────┘

- Вывод основной информации (информация о типах данных, число пропусков, средние значения и т.д)

In [51]:
df_amirov.describe()

shape: (9, 13)
┌────────────┬─────────────┬──────────┬──────────┬───┬───────────┬───────────┬───────┬──────────┐
│ statistic  ┆ PassengerId ┆ Survived ┆ Pclass   ┆ … ┆ Ticket    ┆ Fare      ┆ Cabin ┆ Embarked │
│ ---        ┆ ---         ┆ ---      ┆ ---      ┆   ┆ ---       ┆ ---       ┆ ---   ┆ ---      │
│ str        ┆ f64         ┆ f64      ┆ f64      ┆   ┆ str       ┆ f64       ┆ str   ┆ str      │
╞════════════╪═════════════╪══════════╪══════════╪═══╪═══════════╪═══════════╪═══════╪══════════╡
│ count      ┆ 891.0       ┆ 891.0    ┆ 891.0    ┆ … ┆ 891       ┆ 891.0     ┆ 204   ┆ 889      │
│ null_count ┆ 0.0         ┆ 0.0      ┆ 0.0      ┆ … ┆ 0         ┆ 0.0       ┆ 687   ┆ 2        │
│ mean       ┆ 446.0       ┆ 0.383838 ┆ 2.308642 ┆ … ┆ null      ┆ 32.204208 ┆ null  ┆ null     │
│ std        ┆ 257.353842  ┆ 0.486592 ┆ 0.836071 ┆ … ┆ null      ┆ 49.693429 ┆ null  ┆ null     │
│ min        ┆ 1.0         ┆ 0.0      ┆ 1.0      ┆ … ┆ 110152    ┆ 0.0       ┆ A10   ┆ C        │
│ 25%        ┆ 224.0       ┆ 0.0      ┆ 2.0      ┆ … ┆ null      ┆ 7.925     ┆ null  ┆ null     │
│ 50%        ┆ 446.0       ┆ 0.0      ┆ 3.0      ┆ … ┆ null      ┆ 14.4542   ┆ null  ┆ null     │
│ 75%        ┆ 669.0       ┆ 1.0      ┆ 3.0      ┆ … ┆ null      ┆ 31.0      ┆ null  ┆ null     │
│ max        ┆ 891.0       ┆ 1.0      ┆ 3.0      ┆ … ┆ WE/P 5735 ┆ 512.3292  ┆ T     ┆ S        │
└────────────┴─────────────┴──────────┴──────────┴───┴───────────┴───────────┴───────┴──────────┘

- Количество пассажиров каждого класса

In [52]:
class_count = df_pl.group_by('Pclass').agg(pl.count('Pclass').alias('Count'))
class_count.sort('Pclass')

shape: (3, 2)
┌────────┬───────┐
│ Pclass ┆ Count │
│ ---    ┆ ---   │
│ i64    ┆ u32   │
╞════════╪═══════╡
│ 1      ┆ 216   │
│ 2      ┆ 184   │
│ 3      ┆ 491   │
└────────┴───────┘

- Количество выживших мужчин и женщин на корабле

In [53]:
survived_count = df_pl.filter(pl.col('Survived') == 1).group_by('Sex').agg(pl.count('Survived').alias('Survived_Count'))
survived_count

shape: (2, 2)
┌────────┬────────────────┐
│ Sex    ┆ Survived_Count │
│ ---    ┆ ---            │
│ str    ┆ u32            │
╞════════╪════════════════╡
│ male   ┆ 109            │
│ female ┆ 233            │
└────────┴────────────────┘

- Пассажиры, возраст которых больше 44 лет

In [54]:
older_44 = df_amirov.filter(pl.col('Age') > 44)
older_44.head(5).sort('Age')

shape: (5, 12)
┌─────────────┬──────────┬────────┬──────────────────┬───┬────────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name             ┆ … ┆ Ticket     ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---              ┆   ┆ ---        ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str              ┆   ┆ str        ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪══════════════════╪═══╪════════════╪═════════╪═══════╪══════════╡
│ 53          ┆ 1        ┆ 1      ┆ Harper, Mrs.     ┆ … ┆ PC 17572   ┆ 76.7292 ┆ D33   ┆ C        │
│             ┆          ┆        ┆ Henry Sleeper    ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ (Myna…           ┆   ┆            ┆         ┆       ┆          │
│ 7           ┆ 0        ┆ 1      ┆ McCarthy, Mr.    ┆ … ┆ 17463      ┆ 51.8625 ┆ E46   ┆ S        │
│             ┆          ┆        ┆ Timothy J        ┆   ┆            ┆         ┆       ┆          │
│ 16          ┆ 1        ┆ 2      ┆ Hewlett, Mrs.    ┆ … ┆ 248706     ┆ 16.0    ┆ null  ┆ S        │
│             ┆          ┆        ┆ (Mary D          ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ Kingcome)        ┆   ┆            ┆         ┆       ┆          │
│ 12          ┆ 1        ┆ 1      ┆ Bonnell, Miss.   ┆ … ┆ 113783     ┆ 26.55   ┆ C103  ┆ S        │
│             ┆          ┆        ┆ Elizabeth        ┆   ┆            ┆         ┆       ┆          │
│ 34          ┆ 0        ┆ 2      ┆ Wheadon, Mr.     ┆ … ┆ C.A. 24579 ┆ 10.5    ┆ null  ┆ S        │
│             ┆          ┆        ┆ Edward H         ┆   ┆            ┆         ┆       ┆          │
└─────────────┴──────────┴────────┴──────────────────┴───┴────────────┴─────────┴───────┴──────────┘

# Ускорение работы с Pandas

- Чтение датасета с помощью pandas

In [55]:
df_amirov = pd.read_csv("train.csv")

df_amirov.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

- Определение среднего возраста пассажиров и его стандартное отклонение с помощью bottleneck

In [56]:
mean_age = bn.nanmean(df_amirov['Age'])
mean_age

29.69911764705882

In [57]:
std_age = bn.nanstd(df_amirov['Age'])
std_age

14.516321150817317

- Для каждого пассажира умножьте значение столбца Fare на 1.3, если класс его билета - 1 или 2, и на 1.1, если класс его билета - 3. Сохраните результаты как новый столбец Fare_new. (При реализации расчета используйте любые альтернативы методу iterrows)

In [58]:
df_i = df_amirov

In [59]:
%%timeit

for index, row in df_i.iterrows():
    if row['Pclass'] in [1, 2]:
        df_i.at[index, 'Fare_new'] = row['Fare'] * 1.3
    else:
        df_i.at[index, 'Fare_new'] = row['Fare'] * 1.1

23.3 ms ± 1.99 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [60]:
df_i.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Fare_new  
0      0         A/5 21171   7.2500   NaN        S   7.97500  
1      0          PC 17599  71.2833   C85        C  92.66829  
2      0  STON/O2. 3101282   7.9250   NaN        S   8.71750  
3      0            113803  53.1000  C123        S  69.03000  
4      0            373450   8.0500   NaN        S   8.85500

Альтернативный метод apply:

In [61]:
%%timeit

df_amirov['Fare_new'] = df_amirov.apply(lambda row: row['Fare'] * 1.3 if row['Pclass'] in [1, 2] else row['Fare'] * 1.1, axis=1)

3.11 ms ± 18.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [62]:
df_amirov.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Fare_new  
0      0         A/5 21171   7.2500   NaN        S   7.97500  
1      0          PC 17599  71.2833   C85        C  92.66829  
2      0  STON/O2. 3101282   7.9250   NaN        S   8.71750  
3      0            113803  53.1000  C123        S  69.03000  
4      0            373450   8.0500   NaN        S   8.85500

# Оптимизация типов pandas

- Чтение датасета housing с помощью pandas

In [63]:
pd_amirov2 = pd.read_csv("Housing.csv")

pd_amirov2.head(5)

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0  13300000  7420         4          2        3      yes        no       no   
1  12250000  8960         4          4        4      yes        no       no   
2  12250000  9960         3          2        2      yes        no      yes   
3  12215000  7500         4          2        2      yes        no      yes   
4  11410000  7420         4          1        2      yes       yes      yes   

  hotwaterheating airconditioning  parking prefarea furnishingstatus  
0              no             yes        2      yes        furnished  
1              no             yes        3       no        furnished  
2              no              no        2      yes   semi-furnished  
3              no             yes        3      yes        furnished  
4              no             yes        2       no        furnished

- Для каждого столбца определите оптимальный с точки зрения потребления памяти тип данных - напишите свои выводы в комментариях

In [64]:
pd_amirov2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


In [65]:
pd_amirov2.describe()

price          area    bedrooms   bathrooms     stories  \
count  5.450000e+02    545.000000  545.000000  545.000000  545.000000   
mean   4.766729e+06   5150.541284    2.965138    1.286239    1.805505   
std    1.870440e+06   2170.141023    0.738064    0.502470    0.867492   
min    1.750000e+06   1650.000000    1.000000    1.000000    1.000000   
25%    3.430000e+06   3600.000000    2.000000    1.000000    1.000000   
50%    4.340000e+06   4600.000000    3.000000    1.000000    2.000000   
75%    5.740000e+06   6360.000000    3.000000    2.000000    2.000000   
max    1.330000e+07  16200.000000    6.000000    4.000000    4.000000   

          parking  
count  545.000000  
mean     0.693578  
std      0.861586  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      3.000000

Тип object был преобразован в category, так как столбцы с типом данных object могут содержать категориальную информацию. Данная манипуляция уменьшает объем памяти и улучшает производительность.

Тип int64 был преобразован в int8, так как столбцы с типом данных "int64" содержат целочисленные значения, которые могут быть представлены меньшим количеством бит. Данная манипуляция также позволяет существенно сократить использование памяти.

Тип price и area был преобразован в int32, так как значения в этих столбцах ограничены и не привышают определенных пределов. Это также позволяет сэкономить память без потери точности представления даннных значений.

In [66]:
optimize_types = {
    'price': 'int32',            
    'area': 'int32',             
    'bedrooms': 'int8',          
    'bathrooms': 'int8',         
    'stories': 'int8',           
    'mainroad': 'category',      
    'guestroom': 'category',     
    'basement': 'category',      
    'hotwaterheating': 'category', 
    'airconditioning': 'category', 
    'parking': 'int8',           
    'prefarea': 'category',      
    'furnishingstatus': 'category' 
}

In [67]:
pd_amirov2.astype(optimize_types).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   price             545 non-null    int32   
 1   area              545 non-null    int32   
 2   bedrooms          545 non-null    int8    
 3   bathrooms         545 non-null    int8    
 4   stories           545 non-null    int8    
 5   mainroad          545 non-null    category
 6   guestroom         545 non-null    category
 7   basement          545 non-null    category
 8   hotwaterheating   545 non-null    category
 9   airconditioning   545 non-null    category
 10  parking           545 non-null    int8    
 11  prefarea          545 non-null    category
 12  furnishingstatus  545 non-null    category
dtypes: category(7), int32(2), int8(4)
memory usage: 11.1 KB


Вывод: размер датасета до манипуляций - 55.5 KB, после - 11.1 KB. Благодаря вышеуказанным манипуляциям удалось уменьшить размер датасета в 5 раз.